Add ApplicationPresenter

citizen428 opened the pull request:
@hiattp Took what you did for CR and created a PR here, nice documentation included. Minor semantic changes, see how you like them.

vvmarulin commented:
I think I'm not sure why do we need block. In case when we use only one variable in the view there is no difference if we just wrap this variable in a controller. Also I was thinking about a case when we warp an associated collection and get something like this:
```erb
<div>
  <% article.comments.each do |comment| %>
    <%= CommentPresenter.new(comment) do |presenter| %>
      <div><%= presenter.some_complicated_method %></div>
      <div><%= presenter.another_complicated_method %></div>
    <% end %>
  <% end %>
</div>
``` 
Looks good but usually we have to do something like `@comments = articles.comments.page(params[:page])` in a controller. So we can wrap comments in the presenter right in the controller. If we do this we will remove ruby code from the view what is the main purpose of decorators/presenters.

hiattp commented:
I'm usually inclined to wrap objects in their _decorators_ at the controller level, but I wouldn't wrap an object or a collection in its _presenter_ in a controller. Since presenters are view-specific, I don't think invoking them in the view is an antipattern (the issue isn't getting _all_ ruby out of views, just business logic or other non-view code, whereas presenters strike me as view code—like better versions of helpers). You also might want presenters that are specific to different views for the same object, which again would encourage presenters to be invoked in the view itself.

Note that some of the `complicated_method` definitions you are referencing might actually be in the decorator, so you won't always need to wrap collections/index markup in a presenter. But also note that the block is just an optional convenience and is more intended for a single object. If I had a collection I'd do something more like:

```
  <% article.comments.map{ |c| CommentPresenter.(c) }.each do |comment| %>
      <div><%= comment.some_complicated_method %></div>
      <div><%= comment.another_complicated_method %></div>
  <% end %>
```

And (since I'm not loving that) we might even want a new `present_collection` method on the `ApplicationPresenter` so we can do something more like this (if you buy my "invoke the presenter directly in the view" argument):

```
  <% CommentPresenter.present_collection(@comments).each do |comment| %>
      <div><%= comment.some_complicated_method %></div>
      <div><%= comment.another_complicated_method %></div>
  <% end %>
```

And in the `ApplicationPresenter`:

```
def self.present_collection(collection)
  collection.map{ |object| self.(object) }
end
```

Thoughts?

citizen428 commented:
@hiattp @vvmarulin 

1. Switched to call and changed the examples in the documentation to use `.()`. 
2. Added documentation that it's ok to use presenter in the view.
3. Added `present_collection` class method. It passed a provided block through, so something like this works:
    ```ruby
        FooPresenter.present_collection([foo1, foo2]) do |foo_presenter|
          puts foo.bar
       end  
    ```
4. Added specs.
